Replace window timer with injected timer

Abban opened the pull request:
Components that use timers are hard to test. This
removed the window timers and replaces them with
an interface that we can mock.

Ticket: https://phabricator.wikimedia.org/T378589

Abban committed:
Replace window timer with injected timer

Components that use timers are hard to test. This
removed the window timers and replaces them with
an interface that we can mock.

Updates all tests and banners to use the new timer.

Ticket: https://phabricator.wikimedia.org/T378589

gbirke commented on test/components/BannerConductor/StateMachine/states/PendingState.spec.ts:
```diff
@@ -19,26 +21,36 @@ describe( 'PendingState', function () {
+			delay = after;
+			fn();
+			return 0;
+		};
```
I'd prefer the `TimerSpy` to collect that information in the implementation (even if it's only needed in this test).

gbirke reviewed (approved):
This is a great change! Moving global functions with side effects to dependencies is looking very good.

gbirke commented on banners/desktop/C24_WMDE_Desktop_DE_00/banner_ctrl.ts:
```diff
@@ -27,6 +27,7 @@ import { LocaleFactoryDe } from '@src/utils/LocaleFactory/LocaleFactoryDe';
 import { createFormItems } from './form_items';
 import { createFormActions } from '@src/createFormActions';
 import { currentCampaignTimePercentage } from './currentCampaignTimePercentage';
+import { WindowTimer } from '@src/utils/Timer';
```
Could you move these imports up in all banners (both `WindowTimer` and `currentCampaignTimePercentage`, to avoid misleading comments?

Abban replied:
I think I'd rather remove the comments, the IDE auto imports are always gonna mess this up

moiikana reviewed (changes requested):
WPDE desktop and mobile seem broken for me here,
as they use the BannerConductor they probably need to `provide` the Timer in their entrypoints too

Abban commented:
Good catch, thanks.

gbirke commented on src/components/BannerConductor/BannerConductor.vue:
```diff
@@ -77,6 +79,8 @@ async function closeHandler( closeEvent: TrackingEvent<void> ): Promise<any> {
 	await stateMachine.changeState( stateFactory.newClosedState( closeEvent ) );
 }
 
+onUnmounted( timer.clearAll );
```
The Banner Conductor is never unmounted, currently the whole code stays in the DOM of the page, just hidden. All the timers continue to tick, leaking memory.

I can see three ways moving forward:
1. Make all components responsible to stop the timers they started (like `KeenSlider`). this can be confusing and hard, esp. if they are not unmounted but just hidden but the fact that they no longer displayed needs to reach them. Worst option IMHO.
2. Pass timer as a dependency to the "final" states of the state machine (closed & not shown) and call `timer.clearAll` there.
3. Add code to the "final" states of the state machines that allows to call [`app.unmount()`](https://vuejs.org/api/application.html#app-unmount) to thoroughly remove the banner and clear up its resources. Since [getting the app instance from a component is being removed from Vue](https://stackoverflow.com/q/75446100/130121), the best way to do this is probably via another injection, e.g. `app.provide( 'unmountBanner', () => app.unmount() )`

Abban replied:
The timers should be cleared in the states anyway, as that's where the business logic for the conductor should be. I don't know why I put it into the components.

We should chat about cleaning up the banner properly on close though, as it would free up some browser memory

gbirke commented on src/components/SoftClose/SoftClose.vue:
```diff
@@ -58,33 +59,26 @@ const props = withDefaults( defineProps<Props>(), {
-onUnmounted( () => {
-	window.clearInterval( timer.value );
-} );
-
```
I was baffled why you removed the `onUnmounted`, but then discovered the whole "BannerConductor is supposed to clear all timers but doesn't" issue.

gbirke reviewed (changes requested):
Wow, this update uncovered some flaws in our banner code, thanks again for doing this improvement! It'll benefit us and the donors in the future

gbirke commented:
I just discovered that there is one more change needed - injecting the timer in the WPDE banners

Abban committed:
Remove import comments in entry points

We previously tried to separate the imports in the
banner entry points into logical sections. This didn't
work well with IDE auto adding them at the bottom each
time a new object was added. It was also a little
confusing as to what should go where.

This removes the comments and embraces js import chaos.

Ticket: https://phabricator.wikimedia.org/T378589
